NIFI-6380: Introduced the notion of Parameters and Parameter Contexts…

markap14 opened the pull request:
… to the code base.

- Added nifi-parameter module
- Added Parameter to nifi-api
- Added ParameterContext and ParameterLookup
- Updated EL to support Parameters
- Updated backend so that any property can make use of Parameters
- Added web endpoint for Parameter Contexts
- Updated Templates to make use of Parameter Contexts
- Updated Versioned Flows to make use of Parameter Contexts
- Updated Stateless NiFi to support Parameters and take Parameters as part of the configuration, not Variables.

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

bbende commented:
Will review...

ottobackwards commented:
That is a huge pr for having an empty jira created about 10 minutes before the pr

markap14 commented:
I work fast :)

ottobackwards commented:
are you on slack?

joewitt commented:
The implication about the JIRA and PR speed is problematic.

If you look at the JIRA it is a child [1] of a parent [2] which links to a well written and detailed feature proposal laying out the full context [3].

[1] https://issues.apache.org/jira/browse/NIFI-6380
[2] https://issues.apache.org/jira/browse/NIFI-6276
[3] https://cwiki.apache.org/confluence/display/NIFI/Better+Support+for+Parameterizing+Flows

ottobackwards commented:
Thanks for pointing in the right direction @joewitt

bbende commented:
@markap14 trying to run a full build I seem to consistently get these test failures:
```
[ERROR] Failures:
[ERROR]   TestStandardParameterContext.testChangingParameterForEnabledControllerService:204 Was able to remove parameter being referenced by Controller Service that is DISABLING
[ERROR]   TestStandardParameterContext.testChangingParameterForRunningProcessor:162 Was able to remove parameter while referencing processor was running
[ERROR]   TestStandardParameterContext.testUpdatesApply:69
```
Also some checkstyle violations in `nifi-expression-language` for unused imports.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1147 @@
+    @ApiOperation(
+        value = "Deletes the Update Request with the given ID",
+        response = ParameterContextUpdateRequestEntity.class,
+        notes = "Deletes the Update Request with the given ID. After a request is created via a POST to /nifi-api/parameter-contexts, it is expected "
```
Good catch.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1147 @@
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("validation-requests")
+    @ApiOperation(
+        value = "Initiate a Validation Request to determine how the validity of components will change if a Paramter Context were to be updated",
```
Good catch.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1147 @@
+        entity.setParameterContexts(parameterContexts);
+
+        // generate the response
+        return generateOkResponse(entity).build();
```
Yes, we do.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1147 @@
+                // Verify READ and WRITE permissions for user, for every component that is affected
+                for (final AffectedComponentEntity entity : affectedComponents) {
+                    final AffectedComponentDTO dto = entity.getComponent();
+                    if (AffectedComponentDTO.COMPONENT_TYPE_PROCESSOR.equals(dto.getReferenceType())) {
```
Yeah, good catch, I seem to have skipped a step.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1147 @@
+            requestRevision,
+            lookup -> {
+                final Authorizable parameterContext = lookup.getParameterContext(parameterContextId);
+                parameterContext.authorize(authorizer, RequestAction.WRITE, NiFiUserUtils.getNiFiUser());
```
Will review this, but I think it may have been an oversight.

markap14 replied:
Yeah, we do need to check that, and also the READ policy on the Parameter Context itself.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/dao/impl/StandardParameterContextDAO.java:
```diff
@@ -0,0 +1,231 @@
+                .anyMatch(paramContext -> paramContext.getName().equals(parameterContextDto.getName()));
+
+            if (conflict) {
+                throw new IllegalStateException("Cannot update Paramter Context name because another Parameter Context already exists with the name '" + parameterContextDto.getName() + "'");
```
Have, i searched the codebase, and you flagged all references to 'Paramter' - thanks

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/dao/impl/StandardParameterContextDAO.java:
```diff
@@ -0,0 +1,231 @@
+
+        if (parameterContextDto.getName() != null) {
+            final boolean conflict = flowManager.getParameterContextManager().getParameterContexts().stream()
+                .anyMatch(paramContext -> paramContext.getName().equals(parameterContextDto.getName()));
```
True. I think the following Filter should be sufficient, then: `.anyMatch(paramContext -> paramContext.getName().equals(parameterContextDto.getName()) && !paramContext.getIdentifier().equals(parameterContextDto.getId()))` Agreed?

bbende replied:
Yup makes sense

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1147 @@
+
+        if (contextDto.getId() == null) {
+            throw new IllegalArgumentException("Parameter Context's ID must be specified");
+        }
```
Sure can add that check.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1147 @@
+    public Response deleteParameterContext(
+        @ApiParam(
+            value = "The version is used to verify the client is working with the latest version of the flow.",
+            required = false)
```
We mark the version and client id as not required, but the request must specify at least one of them. You can include a Client ID and no version, or a version and no Client ID, or both. This is the same pattern followed by all DELETE requests.

markap14 replied:
I suppose it may make sense to clarify that error message...

scottyaslan commented:
Can we add a description field to the parameter context?

scottyaslan commented:
Does it make sense to add auditing history to parameter contexts?

markap14 commented:
@scottyaslan adding a description field to the parameter context makes sense to me. Will add that in. I created a new sub-task (NIFI-6384) to capture the auditing information. It's definitely a good call, but I feel like that should probably not hold up the PR.

markap14 commented:
OK I rebased against master and pushed 2 new commits. The first addresses all of your feedback thus far, I believe, @bbende. The second addresses your comment about parameter context description field @scottyaslan. Thanks!

mcgilman commented:
Will also take a look...

scottyaslan commented:
I am a +1 at this point. Thanks for the contribution @markap14 !

scottyaslan commented:
How do we know if a parameter can be written based on the permissions of the components that reference them? Does a user have write privileges to modify a parameter referenced by a processor that the user does not have permission to edit???

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/parameter/StandardParameterContext.java:
```diff
@@ -0,0 +1,279 @@
+
+            for (final Parameter parameter : updatedParameters) {
+                if (parameter.getValue() == null) {
+                    parameters.remove(parameter.getDescriptor());
```
When a user opts to update a sensitive parameter, the front end does not have the value to re-submit. As a result, the request would be mis-interrupted as a request to delete instead of a request to update the description. What are your thoughts on only considering the incoming request as a delete request when the description is not specified? The name and the sensitive flag are technically immutable so these could be considered optional. If the description is specified then update the parameter. If the parameter is sensitive, then only update the value if the value is non-null in the incoming request. I believe this convention is something that we could easily document in the annotations in the DTO and would not be ambiguous.

markap14 replied:
That's a good call. I had thought about the lack of the sensitive property value on the client side but hadn't come up with a smart way of handling it. I like this approach.

kevdoran replied:
Alternatively, as we expect at least the sensitive flag to always be populated, do we want to say that the front end should pass a fully specified object for parameter creation or update, and that just passing the parameter name with no other field indicates deletion?

That is...

**Indicate update value of param1:**
```
{
    "name": "param1"
    "sensitive": "true"
    "description": "My existing, unchanged description."
    "value": "newSecretValue"
}
```

**Indicate update description of param1 (don't modify value):**
```
{
    "name": "param1"
    "sensitive": "true"
    "description": "My updated description"
}
```

**Indicate delete param1:**
```
{
    "name": "param1"
}
```

And sensitivity and name are both immutable, right?

mcgilman replied:
Thanks for commenting @kevdoran. You are correct and I think what you're proposing here will work well.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1185 @@
+        @ApiResponse(code = 404, message = "The specified resource could not be found."),
+        @ApiResponse(code = 409, message = "The request was valid but NiFi was not in the appropriate state to process it. Retrying the same request later may be successful.")
+    })
+    public Response getParameterContextUpdate(@ApiParam("The ID of the Update Request") @PathParam("id") final String updateRequestId) {
```
There is no authorization check on this endpoint. I understand that we do verify that only the user that submitted the request can retrieve it. But since the request is Parameter Context specific, we should be checking access to it. This comment applies to each async endpoint. They all appear to be missing authorization.

markap14 replied:
Agreed.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1185 @@
+            requestEntity,
+            lookup -> {
+                final Authorizable parameterContext = lookup.getParameterContext(requestEntity.getRequest().getParameterContext().getId());
+                parameterContext.authorize(authorizer, RequestAction.READ, NiFiUserUtils.getNiFiUser());
```
My first comment above is not quite correct as the validation details would be pruned if the user did not have permissions. However, I'm thinking we may still want to add them. The intention of the endpoint is to effectively do a `dry run` of updating the parameter context. Lacking these permissions, they will not be able to execute the update anyways.

markap14 replied:
Fair enough. Will update.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/authorization/StandardAuthorizableLookup.java:
```diff
@@ -158,6 +159,18 @@ public Resource getResource() {
+    private static final Authorizable PARAMETER_CONTEXTS_AUTHORIZABLE = new Authorizable() {
+        @Override
+        public Authorizable getParentAuthorizable() {
+            return null;
```
Makes sense. Will do.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1185 @@
+
+                // Verify READ and WRITE permissions for user, for every component that is affected
+                for (final AffectedComponentEntity entity : affectedComponents) {
+                    final AffectedComponentDTO dto = entity.getComponent();
```
Good catch.

mcgilman replied:
Why isn't the new approach like the `getActiveComponentsAffectedByVariableRegistryUpdate` where we directly return the DTO? This way we wouldn't need to deal with the conditionally nulled out component. I think that would be fine in this case given its usage.

mcgilman replied:
Actually, now that I am more closing testing and evaluating I believe my previous suggestion of only needing to authorize against active components was wrong. We take this approach with variables because the variables are referenced by name. Because of this, an EL statement could refer to a variable or a flowfile attribute. What actually resolves will not be known until the flow is actually executed. Because of this, we do not know definitively if EL references a variable or not.

However, with parameters, we have eliminated this ambiguity and we can now authorize against this too. This is what is allowing us to support sensitive parameters. Because of this, we must authorize against all affected components and not just the active ones. Sorry for the misdirection.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1185 @@
+            @ApiResponse(code = 409, message = "The request was valid but NiFi was not in the appropriate state to process it. Retrying the same request later may be successful.")
+        }
+    )
+    public Response getParameterContexts() {
```
Good catch. We should be authorizing against the ParameterContexts Resource.

mcgilman replied:
I think there are two different approaches here. What I was suggesting is slightly different from what you did here. By authorizing against the ParameterContexts Resource, if we wanted to limit access to a single parameter context we would need to explicitly need to override the policies for all other parameter contexts. Any time a new parameter context is added, it's policies would need to be overridden to limit this users access.

My suggestion was to not require permissions to the ParameterContext Resource. If we wanted to give the user access to a single parameter context, we can override the policies for that parameter context only. However, we still need to allow the user to view the listing of parameter contexts. My suggestion was to simply authorize it against `/flow`. It's what we do for Controller Services and Templates. By doing this any user could open the listing in the UI but they would only be able to act on parameter contexts they have permissions to. If we chose to implement it this way, I would suggest moving this method into the `FlowResource` as that is the resource that is driving the access decision.

Alternatively, we could update the logic to check against the ParameterContexts Resource. If the user does not have permissions to it, we could then check the policy for every ParameterContext. If any of these are allowed, then we could short circuit. Additionally, we would need to return some flag that represents this permission so we can conditionally provide the Parameter Context menu item in the global menu.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1185 @@
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("update-requests")
```
Agreed.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java:
```diff
@@ -0,0 +1,1185 @@
+        // 9. Re-Enable all affected Controller Services
+        // 10. Re-Start all Processors
+
+        final Set<AffectedComponentEntity> affectedComponents = serviceFacade.getComponentsAffectedByParameterContextUpdate(contextDto);
```
That's a good catch. Here, we are validating that the user has READ & WRITE permissions to all affected Components, but we only need to validate this for components that are active. Other components won't be affected, so the user doesn't need to have these permissions.

mcgilman replied:
Please see my comment below [1]. I believe my suggestion here was incorrect.

[1] https://github.com/apache/nifi/pull/3536#discussion_r301683230

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/flow/StandardFlowManager.java:
```diff
@@ -582,7 +591,11 @@ public void removeReportingTask(final ReportingTaskNode reportingTaskNode) {
+        // TODO: Probably should create a method on ComponentNode: `List<ControllerServiceNode> getReferencedControllerServices()` This same code block (or something similar)
+        //  is scattered all throughout the codebase. If we look at references to ComponentNode.getEffectivePropertyValues(), it will show the same thing over & over.
+        //  ALSO - double-check what happens if we have a component reference Controller Service A two times - here we would remove the reference twice, but do we add the reference
+        //  twice? If not, we need to address that!
```
Are these things that you still need to do?

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/flow/StandardFlowManager.java:
```diff
@@ -638,7 +651,8 @@ public void removeRootControllerService(final ControllerServiceNode service) {
         }
 
-        for (final Map.Entry<PropertyDescriptor, String> entry : service.getProperties().entrySet()) {
+        // TODO: Again - add getReferencedControllerServices() method. It would be more efficient than this, calculating the effective property values every time.
```
And this?

markap14 commented:
@mcgilman good catch on the comments/TODO's left in the code. Will push another commit addressing these and with a few additional bug fixes around handling Parameter Contexts referenced by components that the user doesn't have access to

mcgilman commented:
@markap14 Thanks for the update. I've left some follow up comments in my previous feedback. A couple of other items:

- Asked earlier but did not see any response... What are your thoughts on automatically setting the parameter context on newly created Process Groups based on the currently bound parameter context in the parent Process Group? Since parameters are not inherited, this could be a convenient feature if users would expect this behavior. There is a downside to consider that changing the bound parameter context would be cumbersome but this downside would exist regardless if this convenience were implemented.
- When updating a Process Group to bind a Parameter Context, I am able to even if I do not have access to that Process Group. I think we need to update this and perform this check.
- When updating a Process Group to bind a Parameter Context, I am able to even if that Parameter Context has affected components that I do not have access to. I think we need to update this and perform these checks.
- When updating a Process Group to unbind a Parameter Context, I am able to even if I do not have access to that Process Group. This check is questionable in my opinion. We do require permissions to a referenced Controller Service when it is being dereferenced. So I could see adding that check here too however with Controller Services they can be affected by the components referencing them. This is not really the case here but I still think it seems reasonable to add.
- When updating a Process Group to unbind a Parameter Context, I am able to even if that Parameter Context has affected components that I do not have access to. I think we need to update this and perform these checks.

markap14 commented:
@mcgilman thanks for the review - again :) I am not opposed to the idea of automatically setting the Parameter Context to the same as the parent's when a Process Group is created. From a User Experience point of view, it's probably a good idea. The only real concern that I have is that in order to do that, we'd have to first receive the request, determine if the user has permissions to the parent group's Parameter Context, and if so, set it on the request before replicating. If any node then determined that the user doesn't have the READ policy for the Parameter Context (of course different nodes could see the permissions differently if they are periodically syncing with an external authority provider), then the result will be that the request is rejected. So the user has all permissions that they should, and the request still gets rejected due to insufficient permissions, because they don't have READ on the parent's Parameter Context. Now, this is certainly a corner case, so if you think we should go ahead and do this anyway I'm happy to make the change. But wanted to just run that scenario by you and get your feedback on that first.

Regarding permissions of Binding & Un-binding: you're right - the current approach is inconsistent with how it works with Controller Services. Have updated code. Will push a new commit. Good catch!

bbende commented:
Been testing this the past couple of days and had two comments...

1) I think there is some inconsistency with the authorization hierarchy for an individual context (currently has parent of parameter-contexts, with parent of controller) vs the parameter-contexts resource (currently no parent).

2) I tested all of the parameter reference examples from the feature proposal... 

When using a property that does not support EL, the results matched the expected results on the wiki.

When using a property that does support EL, there were a few weird results related to escaping.

Using a context with a parameter name `param1` having a value of `param1-dev`:

```
###{param1} -> ###param1-dev (should be #param1-dev)
####{param1} -> ####{param1}####{param1}  (should be ##{param1})
#####{param1} -> #####param1-dev (should be ##param1-dev)
```

markap14 commented:
Thanks @bbende  - pushed a new commit that addresses the above. Added integration tests to verify behavior is consistent with those outlined above, regardless of whether or not property descriptor supports EL.

bbende commented:
Re-testing the references looks like two of them are still slightly off when the property supports EL...

```
##{param1} -> param1-dev (should be #{param1})
####{param1} -> #{param1} (should be ##{param1})
```

markap14 commented:
OK I pushed an update to the PR. However, since we are changing the comparison logic, that has to be done in the nifi-registry project, so this is now blocked by NIFIREG-294.

markap14 commented:
@bbende I see. That appears to be due to the way that UpdateAttribute evaluates the Parameters. Will have a new commit up shortly to address. Thanks!

bbende commented:
The parameter references look good with the latest commit, also tested all the standard operations in a secure instance and looks good. I'm a +1.

mcgilman commented:
@markap14 Things are looking really good here. I am seeing one minor issue still.

For Reporting Tasks and Controller Services in the Settings cannot reference parameters. However, when I attempt reference one the validation error recognizes the parameter and states that 'no Parameter Context is bound to this Process Group'.

markap14 commented:
Thanks @mcgilman I pushed an update to the validation error text.

mcgilman commented:
Thanks @markap14! This has been merged to master.

markap14 commented:
Thanks @mcgilman for all the effort you put into reviewing this and all the feedback! Definitely excited to have this feature added in!
